bluetooth: host: CS support for various HCI commands

adamfc2000 opened the pull request:
Adds HCI support for:
- LE CS Security Enable
- LE CS Procedure Enable
- LE CS Set Procedure Parameters
- LE CS Set Channel Classification
- LE CS Read Local Supported Capabilities
- LE CS Write Cached Remote Supported Capabilities
- LE CS Write Cached Remote FAE Table

olivier-le-sage commented on include/zephyr/bluetooth/cs.h:
```diff
@@ -620,6 +620,172 @@ int bt_le_cs_create_config(struct bt_conn *conn, struct bt_le_cs_create_config_p
+ * 
+ * @note To use this API, @kconfig{CONFIG_BT_CHANNEL_SOUNDING} must be set.
+ * 
+ * @param channel_classification Bit fields
```
This could maybe document the requirements for the channel map and/or reference bt_le_cs_set_valid_chmap_bits

olivier-le-sage commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -2402,12 +2402,70 @@ struct bt_hci_cp_le_tx_test_v4_tx_power {
 
+#define BT_HCI_OP_LE_CS_READ_LOCAL_SUPPORTED_CAPABILITIES BT_OP(BT_OGF_LE, 0x0089) /* 0x2089 */
+
+struct bt_hci_rp_le_read_local_supported_capabilities {
```
rp?

jhedberg replied:
> rp?

response parameters, a structure naming convention used throughout this file

olivier-le-sage replied:
ah ok. I missed that it was for the event. We've moved this back to draft btw

olivier-le-sage commented on subsys/bluetooth/host/cs.c:
```diff
@@ -447,4 +447,197 @@ int bt_le_cs_remove_config(struct bt_conn *conn, uint8_t config_id)
+
+	cp = net_buf_add(buf, 10);
+
+	if (sys_cpu_to_le16(0x0001) == 0x0001)
```
I think you can just memcpy(). That's what we did here https://github.com/zephyrproject-rtos/zephyr/blob/4260f07a468aee09165d142a014c7759a55805d8/subsys/bluetooth/host/cs.c#L429

It does mean the input to this function needs to be HCI-formatted

olivier-le-sage commented on subsys/bluetooth/host/cs.c:
```diff
@@ -447,4 +447,197 @@ int bt_le_cs_remove_config(struct bt_conn *conn, uint8_t config_id)
+		return err;
+	}
+
+	BTA_EXPECT(rsp != NULL);
```
guessing BTA_EXPECT here and below isn't intentional ?

olivier-le-sage commented on subsys/bluetooth/host/cs.c:
```diff
@@ -447,4 +447,197 @@ int bt_le_cs_remove_config(struct bt_conn *conn, uint8_t config_id)
+
+	cp->handle = sys_cpu_to_le16(conn->handle);
+
+	for (int i = 0; i < sizeof(cp->remote_fae_table); i++) {
```
can simply memcpy()

olivier-le-sage commented on subsys/bluetooth/host/cs.c:
```diff
@@ -447,4 +447,197 @@ int bt_le_cs_remove_config(struct bt_conn *conn, uint8_t config_id)
+int bt_le_cs_write_cached_remote_supported_capabilities(struct bt_conn *conn, const struct bt_le_cs_write_cached_remote_supported_capabilities_param *params)
+{
+	struct bt_hci_cp_le_write_cached_remote_supported_capabilities *cp;
+	struct net_buff *buf;
```
net_buf

olivier-le-sage commented on include/zephyr/bluetooth/cs.h:
```diff
@@ -620,6 +620,172 @@ int bt_le_cs_create_config(struct bt_conn *conn, struct bt_le_cs_create_config_p
+ */
+int bt_le_cs_set_channel_classification(uint8_t channel_classification[10]);
+
+struct bt_le_cs_read_local_supported_capabilities_return {
```
So I was thinking we wouldn't need:
- struct bt_le_cs_read_local_supported_capabilities_return
- struct bt_le_cs_write_cached_remote_supported_capabilities_param

and could just reuse struct bt_conn_le_cs_capabilities instead

olivier-le-sage commented on include/zephyr/bluetooth/cs.h:
```diff
@@ -620,6 +620,172 @@ int bt_le_cs_create_config(struct bt_conn *conn, struct bt_le_cs_create_config_p
+struct bt_le_cs_procedure_enable_param {
+	uint8_t config_id;
+	uint8_t enable;
+}
```
missing semicolon ?

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1396,6 +1396,20 @@ enum bt_security_err {
 	BT_SECURITY_ERR_UNSPECIFIED,
 };
 
+struct bt_conn_le_cs_procedure_enable_complete {
```
For this file and cs.h I think it would be good to document what each parameter means and what values are allowed.
Otherwise an app developer needs to know the core spec/have the core spec open to be able to use this

olivier-le-sage commented on include/zephyr/bluetooth/cs.h:
```diff
@@ -620,6 +620,136 @@ int bt_le_cs_create_config(struct bt_conn *conn, struct bt_le_cs_create_config_p
+	uint16_t max_procedure_count;
+	uint32_t min_subevent_len;
+	uint32_t max_subevent_len;
+	uint8_t tone_antenna_config_selection;
```
I think using enums helps prevent users from using wrong values

For 
- tone_antenna_config_selection
- snr_control_initiator
- snr_control_reflector

you can re-use the existing enums in this file. It looks like they just need to be renamed to remove the word "test"

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1396,6 +1396,20 @@ enum bt_security_err {
 
+struct bt_conn_le_cs_procedure_enable_complete {
+	uint8_t config_id;
+	uint8_t state;
```
can we have a simple enum for state? Then re-use it here https://github.com/zephyrproject-rtos/zephyr/pull/79183/files#diff-b8bf0043e1d142bd1bc51849856bc2bf6b8e905f25d7bdba9a3d2d934547cce5R639

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1677,6 +1691,27 @@ struct bt_conn_cb {
+	 *  @param params CS Procedure Enable parameters
+	 */
+	void (*le_cs_procedure_enable_available)(struct bt_conn *conn,
+				struct bt_conn_le_cs_procedure_enable_complete *params);
```
btw the convention in these files is apparently 1 tab = 8 spaces

olivier-le-sage commented on include/zephyr/bluetooth/cs.h:
```diff
@@ -620,6 +620,136 @@ int bt_le_cs_create_config(struct bt_conn *conn, struct bt_le_cs_create_config_p
+	uint32_t min_subevent_len;
+	uint32_t max_subevent_len;
+	uint8_t tone_antenna_config_selection;
+	uint8_t phy;
```
maybe expose an enum for the PHY. At first I thought this included 2BT 2M but seems not

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1677,6 +1691,27 @@ struct bt_conn_cb {
+	 *  @param conn Connection object.
+	 *  @param params CS Procedure Enable parameters
+	 */
+	void (*le_cs_procedure_enable_available)(struct bt_conn *conn,
```
```suggestion
	void (*le_cs_procedure_enabled)(struct bt_conn *conn,
```

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1677,6 +1691,27 @@ struct bt_conn_cb {
+	 */
+	void (*le_cs_security_enable_available)(struct bt_conn *conn);
+
+	/** @brief LE CS Procedure Enable Complete Event.
```
```suggestion
	/** @brief LE CS Procedure Enabled.
```

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1677,6 +1691,27 @@ struct bt_conn_cb {
 	 */
 	void (*le_cs_config_removed)(struct bt_conn *conn, uint8_t config_id);
+
+	/** @brief LE CS Security Enable Complete Event.
```
```suggestion
	/** @brief LE CS Security Enabled
```

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1677,6 +1691,27 @@ struct bt_conn_cb {
+	 *
+	 *  @param conn Connection object.
+	 */
+	void (*le_cs_security_enable_available)(struct bt_conn *conn);
```
```suggestion
	void (*le_cs_security_enable_available)(struct bt_conn *conn, uint8_t status);
```
?

adamfc2000 committed:
bluetooth: host: CS support for various HCI commands

Adds HCI support for:
- LE CS Security Enable
- LE CS Procedure Enable
- LE CS Set Procedure Parameters
- LE CS Set Channel Classification
- LE CS Read Local Supported Capabilities
- LE CS Write Cached Remote Supported Capabilities
- LE CS Write Cached Remote FAE Table

Signed-off-by: Adam Cavender <adam.cavender@nordicsemi.no>

theob-pro commented on include/zephyr/bluetooth/cs.h:
```diff
@@ -687,6 +649,178 @@ void bt_le_cs_step_data_parse(struct net_buf_simple *step_data_buf,
+	enum bt_le_cs_procedure_phy phy;
+
+	/* Transmit power delta, in signed dB, to indicate the recommended difference between the
+	 * remote device's power level for the CS tomes and RTT packets and the existing power
```
```suggestion
	 * remote device's power level for the CS tones and RTT packets and the existing power
```

olivier-le-sage commented on subsys/bluetooth/host/shell/cs.c:
```diff
@@ -442,6 +442,255 @@ static int cmd_cs_stop_test(const struct shell *sh, size_t argc, char *argv[])
+	params.tx_power_delta = 0x80;
+	params.preferred_peer_antenna = 1;
+	params.snr_control_initiator = BT_LE_CS_INITIATOR_SNR_CONTROL_18dB;
+	params.snr_control_initiator = BT_HCI_OP_LE_CS_REFLECTOR_SNR_18;
```
you probably intended to write to `snr_control_reflector`
